src/gt4py/next/iterator/transforms/global_tmps.py

```diff
@@ -8,582 +8,198 @@
-    ] = None,
-) -> FencilWithTemporaries:
-    """Split closures on lifted function calls and introduce new temporary buffers for return values.
-
-    Newly introduced temporaries will have the symbolic size of `AUTO_DOMAIN`. A symbol with the
-    same name is also added as a fencil argument (to be replaced at a later stage).
-
-    For each closure, follows these steps:
-    1. Pops up lifted function calls to the top of the expression tree.
-    2. Introduce new temporary for the output.
-    3. Extract lifted function class as new closures with the previously created temporary as output.
-    The closures are processed in reverse order to properly respect the dependencies.
-    """
-    if not extraction_heuristics:
-        # extract all (eligible) lifts
-        def always_extract_heuristics(_: ir.StencilClosure) -> Callable[[ir.Expr], bool]:
-            return lambda _: True
-
-        extraction_heuristics = always_extract_heuristics
-
-    uid_gen_tmps = UIDGenerator(prefix="_tmp")
-
-    node = itir_type_inference.infer(node, offset_provider=offset_provider)
-
-    tmps: list[tuple[str, ts.DataType]] = []
-
-    closures: list[ir.StencilClosure] = []
-    for closure in reversed(node.closures):
-        closure_stack: list[ir.StencilClosure] = [closure]
-        while closure_stack:
-            current_closure: ir.StencilClosure = closure_stack.pop()
-
-            if (
-                isinstance(current_closure.stencil, ir.SymRef)
-                and current_closure.stencil.id == "deref"
-            ):
-                closures.append(current_closure)
-                continue
-
-            is_scan: bool = cpm.is_call_to(current_closure.stencil, "scan")
-            current_closure_stencil = (
-                current_closure.stencil if not is_scan else current_closure.stencil.args[0]  # type: ignore[attr-defined]  # ensured by is_scan
+            tmp_dtypes: ts.ScalarType | tuple[ts.ScalarType | tuple, ...] = (
+                type_info.apply_to_primitive_constituents(
+                    type_info.extract_dtype,
+                    tmp_expr.type,
+                    tuple_constructor=lambda *elements: tuple(elements),
+                )
             )
 
-            extraction_predicate = TemporaryExtractionPredicate(
-                extraction_heuristics(current_closure)
-            )
+            # allocate temporary for all tuple elements
+            def allocate_temporary(tmp_name: str, dtype: ts.ScalarType, domain: itir.Expr):
+                declarations.append(itir.Temporary(id=tmp_name, domain=domain, dtype=dtype))
 
-            stencil_body, extracted_lifts, _ = extract_subexpression(
-                current_closure_stencil.expr,
-                extraction_predicate,
-                uid_gen_tmps,
-                once_only=True,
-                deepest_expr_first=True,
+            next_utils.tree_map(functools.partial(allocate_temporary, domain=domain_expr))(
+                tmp_names, tmp_dtypes
```

Comment:
The declarations array is not very important for understanding the pass and returning it would severely impede the readability of the rest because one could not execute `_transform_stmt` inline anymore, but instead would need to extract it's results and continuously merge the declarations. Take a look at `_transform_if` and how it would look like in that case.